src/team_lib/multirail/xccl_mrail_lib.c

```diff
@@ -284,11 +355,16 @@ xccl_status_t xccl_mrail_team_destroy(xccl_tl_team_t *team)
 xccl_status_t xccl_mrail_context_progress(xccl_tl_context_t *team_context)
 {
     int                  i;
-    xccl_mrail_context_t *mrail_ctx = ucs_derived_of(team_context,
+    xccl_mrail_context_t  *mrail_ctx = ucs_derived_of(team_context,
                                                      xccl_mrail_context_t);
+    xccl_team_lib_mrail_t *mrail     = ucs_derived_of(team_context->lib,
+                                                     xccl_team_lib_mrail_t);
 
-    for(i = 0; i < mrail_ctx->n_tls; i++) {
-        xccl_context_progress(mrail_ctx->tls[i]);
+    // TODO: if there are no progress threads then progress here 
+    if (mrail->config.threads_num == 0) {
```

Comment:
btw, another idea: if threads_num > 0 you can modify the static lib structure and set .super.team_context_progress to NULL. We already have this branch at xccl/core. Then you don't need this branch here.